gpu/octree/src/cuda/octree_host.cu

```diff
@@ -100,6 +100,57 @@ namespace
         return count;
     } 
 
+    int nearestVoxelTraversal(float3 query, int level, int mask, float3 minp, float3 maxp, int& x, int& y, int& z)
+    {
+        //identify closest voxel
+        float closest_distance = std::numeric_limits<float>::max();
+        int closest_index = 0, closest_x = 0, closest_y = 0, closest_z = 0;
+
+        for (int i = 0; i < 8; ++i)
+        {
+            if ((mask & (1<<i)) == 0)   //no child
+                continue;
+
+            //calculate  x,y,z offset for voxel
+            int x_cord = i & 1;
+            int y_cord = (i>>1) & 1;
+            int z_cord = (i>>2) & 1;
+
+            int x_child, y_child, z_child;
+            x_child = x*2 + x_cord;
+            y_child = y*2 + y_cord;
+            z_child = z*2 + z_cord;
+
+            //find center of child cell
+            float3 voxel_center;
+            voxel_center.x = minp.x + (maxp.x - minp.x) * (2*x_child + 1) / (2 * 1<<(level + 1));
+            voxel_center.y = minp.y + (maxp.y - minp.y) * (2*y_child + 1) / (2 * 1<<(level + 1));
+            voxel_center.z = minp.z + (maxp.z - minp.z) * (2*z_child + 1) / (2 * 1<<(level + 1));
+
+            //compute distance to centroid
+            float dx = (voxel_center.x - query.x);
+            float dy = (voxel_center.y - query.y);
+            float dz = (voxel_center.z - query.z);
+            float distance_to_query = dx * dx + dy * dy + dz * dz;
+
+            //compare distance
+            if (distance_to_query < closest_distance)
+            {
+                closest_distance = distance_to_query;
+                closest_index = i;
+                closest_x = x_child;
+                closest_y = y_child;
+                closest_z = z_child;
+            }
+        }
+
+        x = closest_x;
+        y = closest_y;
+        z = closest_z;
```

Comment:
You're not making use of this outside this function.

Reply:
Since I've passed `x,y,z` by reference, these are updated in each traversal step, and the updated values are used in the next traversal step. They're only unused at the last iteration of the traversal.